### PR TITLE
Update Composition Functions content to v1beta1

### PR DESCRIPTION
**Pull request by negz**
Relates to https://github.com/crossplane/docs/issues/579
Relates to https://github.com/crossplane/docs/issues/582

This PR documents how to use composition functions. It also covers everything I think we need to cover about how they work and how to write them, short of language-specific guides. I plan to add a more detailed language-specific guide for Go that covers function-sdk-go and function-template-go in a separate PR.

### CODE REVIEW

File: content/master/concepts/compositions.md [hunk 1]
```diff
@@ -664,40 +664,59 @@ spec:
 The [EnvironmentConfigs]({{<ref "./environment-configs" >}}) page has 
 more information on EnvironmentConfigs options and usage.
 
-### Composition functions
+### Use Composition Functions
```
Comment (negz) [inline content/master/concepts/compositions.md:667]:
```
 667:5   warning  'Use Composition Functions'     Google.Headings
                  should use sentence-style
                  capitalization.
```

I'm feeling a little all over the place WRT whether I should capitalize Composition Functions, here in the heading and elsewhere. Some context that may help figure out the right pattern:

* Composition is a type. I couldn't find a clear pattern for when we say Composition vs composition in the docs.
* Function is a type (of package).
* I'd like to call Composition Functions just Functions" (or functions?) for short. Not xfns or XFNs.

Comment (negz) [reply content/master/concepts/compositions.md:667]: I mention this because if I'm following https://developers.google.com/style/headings correctly I think Vale would accept "Use composition functions" here as the heading.

Comment (plumbis) [reply content/master/concepts/compositions.md:667]: We've generally tried to follow the [Kubernetes docs guidelines](https://kubernetes.io/docs/contribute/style/style-guide/) on this. 

My approach is if I'm talking about a concept I use sentence case. I (try) and only capitalize when referring to the type, for example "Creating composite resource definitions" vs "Create a CompositeResourceDefinition object"

For titles, _most_ of the time I'd recommend sentence case with lowercase `functions` which may require rewording like "Use a composition function"

**Comment by negz** [comment]
Here's a rough draft of the topics I think I want to cover and where:

In `concepts/composition-functions.md`
- [ ] What a function is
- [ ] Installing a Function package
- [ ] Using functions
  - [ ] Using a function in a Composition
  - [ ] Using a pipeline of functions in a Composition
  - [ ] Test a Composition that uses functions using `crossplane beta render`
  - [ ] Links to interesting functions folks have built
- [ ] Develop a function
  - [ ] How a function works (developer focused)
  - [ ] RunFunctionRequest and RunFunctionResponse overview
  - [ ] Tips and gotchas for writing a Function
  - [ ] Links to more detailed, language specific guides

In `knowledge-base/guides/...`:
- [ ] Detailed, step-by-step guide to writing a function in Go

I think each of these Functions could use some documentation somewhere:

* https://github.com/crossplane-contrib/function-patch-and-transform
* https://github.com/crossplane-contrib/function-go-templating/tree/main
* https://github.com/crossplane-contrib/function-cue
* https://github.com/crossplane-contrib/function-auto-ready

I'm not sure whether that should be the Marketplace or the docs site (knowledge base)?

**Comment by negz** [comment]
@plumbis I'm (mostly) able to generate a Hugo-compatible page of API documentation for the types Crossplane uses to communicate with functions. See https://github.com/crossplane/crossplane/pull/4914. These types are pretty deep - the SDKs will abstract them away a bit - but they're probably worth having access to for advanced function developers.

Would it make sense to include these somewhere under content? Any idea where? They're fairly coupled to a Crossplane release.

**Comment by plumbis** [comment]
Is this a different type of API doc than the Crossplane CRDs?

**Comment by negz** [comment]
@plumbis Yeah, these docs are generated from protobuf schemas rather than CRD OpenAPI schemas. So it’s a different tool doing the generation. (See See https://github.com/crossplane/crossplane/pull/4914.)

File: content/master/concepts/composition-functions.md [hunk 2]
```diff
@@ -1,1026 +1,637 @@
...
 
-* An optional, arbitrary `config` object.
-* The `observed` state of the XR, any existing composed resources, and their connection details.
-* The `desired` state of the XR and any composed resources.
-* Optional `results` of the Function pipeline.
 
-Here's a brief example of a `FunctionIO`:
+<!-- vale write-good.Passive = NO -->
+Here's an example of a tiny, hello world function. This example is written in
+Go.
```
Comment (negz) [inline content/master/concepts/composition-functions.md:315]: @plumbis, Vale seems to be parsing "Go." (with period) as a word and complaining that it's a spelling mistake. Any idea how to fix that? I added Go (no period) to the dictionary.

File: content/master/concepts/composition-functions.md [hunk 3]
```diff
@@ -1,1026 +1,637 @@
...
+```
 
-This means Functions may be written using a general-purpose programming language
-like Python, Go, or TypeScript. They may also be implemented using a shell
-script, or an existing tool like Helm or Kustomize.
+{{<hint "important">}}
+Running `crossplane beta render` requires [Docker](https://www.docker.com).
 
-### FunctionIO
+See the [Crossplane CLI docs](https://github.com/crossplane/docs/pull/584) to
```
Comment (negz) [inline content/master/concepts/composition-functions.md:302]: This link is a placeholder. Needs updating when https://github.com/crossplane/docs/pull/584 is merged.

File: content/master/concepts/composition-functions.md [hunk 4]
```diff
@@ -1,1026 +1,637 @@
...
+1. Edit the template to add your function's logic.
+1. Test your function.
+1. Build your function, and push it to a package registry.
 
-When Crossplane encounters a `Fatal` result it will finish running the
-Composition Function pipeline. Crossplane will then return an error without
-applying any changes to the API server. Crossplane surfaces this error as a
-`Warning` event, a debug log, and by setting the `Synced` status condition of
-the XR to "False".
+You use the [Crossplane CLI](https://github.com/crossplane/docs/pull/584) to
```
Comment (negz) [inline content/master/concepts/composition-functions.md:366]: This link is a placeholder. Needs updating when https://github.com/crossplane/docs/pull/584 is merged.

**Review by negz — COMMENTED** [review]
@plumbis @phisco this is ready for review. There's more documentation to be done, but I can handle it in separate PRs. Namely I want to:

* Add a language-specific guide for Go.
* Make sure we're good on the packages front once https://github.com/crossplane/docs/pull/566 is merged.

File: content/master/concepts/composition-functions.md [hunk 5]
```diff
@@ -1,1026 +1,637 @@
...
+a high performance, open source remote procedure call (RPC) framework. When you
+[install a function](#install-a-composition-function) Crossplane deploys the
+function as a gRPC server. Crossplane encrypts and authenticates all gRPC
+communication.
+
+You don't have to be a gRPC expert to write a function. Crossplane's function
+SDKs setup gRPC for you. It's useful to understand how Crossplane calls your
+function though, and how your function should respond.
+
+```mermaid
```
Comment (negz) [inline content/master/concepts/composition-functions.md:414]: This looks great to me in light mode, but not so much in dark mode. It renders the line labels as dark grey on a black background.

File: content/master/concepts/composition-functions.md [hunk 6]
```diff
@@ -1,1026 +1,699 @@
...
+should add these fields every time Crossplane calls it. If a function adds a
+field to the desired state once, but doesn't add it the next time it's called,
+Crossplane deletes the field. The same is true for composed resources. If a
+function adds a composed resource to the desired state, but doesn't add it the
+next time it's called, Crossplane deletes the composed resource.
 
-Create and push the Function just like you would any Docker image.
+{{<hint "tip">}}
+Crossplane uses
+[server side apply](https://kubernetes.io/docs/reference/using-api/apply/)
```
Comment (TerjeLafton) [inline content/master/concepts/composition-functions.md:597]: This URL gives me a 404.
Was it meant to be this?
https://kubernetes.io/docs/reference/using-api/server-side-apply/

**Comment by TerjeLafton** [comment]
I haven't read very thoroughly, but this seem really nice to me!

File: content/master/concepts/composition-functions.md [hunk 7]
```diff
@@ -1,1026 +1,699 @@
...
+determine what resources it should create when you create a composite resource
+(XR). You can write a function to template resources using a general purpose
+programming language like Go or Python. Using a general purpose programming
+language allows a Function to use more advanced logic to template resources,
+like loops and conditionals.
 
-Composition Functions allow you to supplement or replace your Compositions with
-advanced logic not implementable through available patching strategies. 
+You can build a function using general purpose programming languages such as Go
+or Python. The Crossplane community has also built functions that let you
```
Comment (JonasKs) [inline content/master/concepts/composition-functions.md:21]: nit: felt a bit repetative, since you've already said you can use a general purpose programming language like Go or Python two sentences ago

**Review by JonasKs — COMMENTED** [review]
Looks really good, thanks for this! 😊 
As I said in slack, this image is hard to read when using dark mode: 
image

File: themes/geekboot/layouts/partials/single-list.html [hunk 8]
```diff
@@ -75,3 +75,10 @@ <h1 class="bd-title mb-0" id="content">{{ .Title | markdownify }}</h1>
         </div>
     </main>
   </div>
+
+  {{ if .Page.Store.Get "hasMermaid" }}
+  <script type="module">
+    import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.esm.min.mjs';
+    mermaid.initialize({ startOnLoad: true });
+  </script>
```
Comment (plumbis) [inline themes/geekboot/layouts/partials/single-list.html:83]:
```suggestion
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js" type="text/javascript"></script>
<script type="module">

  import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.esm.min.mjs';
  document.addEventListener("DOMContentLoaded", setMermaidStyle());
  document.addEventListener("DOMContentLoaded", colorModeListener());

  function getMermaidConfig(){
    var style = getComputedStyle(document.body)
    var font = style.getPropertyValue("font-family")
    var fontColor = style.getPropertyValue('--body-font-color')
    var backgroundColor = style.getPropertyValue('--body-background')

    var config = {
      "theme": "base",
      "fontFamily": font,
      "themeVariables": {
        "background": backgroundColor,
        "textColor": fontColor,
        }
    }

    return config
  }

  function setMermaidStyle(){
    var config = getMermaidConfig()
    mermaid.initialize( config )
  }

  function colorModeListener(){
    darkSwitch.addEventListener("click", resetMermaidStyle())
  }

  function resetMermaidStyle(){
    console.log("resetting")

    var config = getMermaidConfig()
    console.log(config)
    mermaid.mermaidAPI.setConfig( config )
    mermaid.mermaidAPI.reset()

  }

</script>
```

This could look better but it's a good start to make the diagrams match the light/dark theme.

There is a bug in here that switching themes doesn't update the diagram (but it works on future loads)

File: content/master/concepts/composition-functions.md [hunk 9]
```diff
@@ -1,1026 +1,699 @@
...
+  This linter does not allow "auto-ready", which is part of the name of the
+  function.
+-->
+Crossplane passes each Function in the pipeline the result of the previous
+Function. This enables powerful combinations of Functions. In this example,
+Crossplane calls {{<hover label="double" line="10">}}function-cue{{</hover>}} to
+create an S3 bucket. Crossplane then passes the bucket to {{<hover
+label="double" line="23">}}function-auto-ready{{</hover>}}, which marks the
+composite resource as ready when the bucket becomes ready.
+<!-- vale Microsoft.Auto = YES -->
```
Comment (plumbis) [inline content/master/concepts/composition-functions.md:167]:
```suggestion
Crossplane passes each Function in the pipeline the result of the previous
Function. This enables powerful combinations of Functions. In this example,
Crossplane calls {{<hover label="double" line="10">}}function-cue{{</hover>}} to
create an S3 bucket. Crossplane then passes the bucket to 
{{<hover label="double" line="23">}}function-auto-ready{{</hover>}}, which marks the
composite resource as ready when the bucket becomes ready.
```

Vale ignores the contents of `{{<hover >}}` elements but it has to be on a single line

File: content/master/concepts/composition-functions.md [hunk 10]
```diff
@@ -1,1026 +1,699 @@
...
+
+The Crossplane CLI uses Docker Engine to run functions. You can change how the
+Crossplane CLI run a function by adding an annotation in `functions.yaml`. Add
+the `render.crossplane.io/runtime` annotation to a Function to change how it's
+run.
+
+`crossplane beta render` supports two `render.crossplane.io/runtime` values:
+
+* `Docker` (the default) connects to Docker Engine. It uses Docker to pull and
+  run function runtimes.
```
Comment (plumbis) [inline content/master/concepts/composition-functions.md:318]:
```suggestion
  run a function runtime.
```

vale fix

File: content/master/concepts/composition-functions.md [hunk 11]
```diff
@@ -1,1026 +1,699 @@
...
+    render.crossplane.io/runtime: Development
+spec:
+  package: xpkg.upbound.io/crossplane-contrib/function-patch-and-transform:v0.1.4
+```
 
-Here's a brief example of a `FunctionIO`:
+{{<hint "tip">}}
+Use the `Development` runtime when you
+[write a composition function](#write-a-composition-function) to test your
+function end-to-end.
```
Comment (plumbis) [inline content/master/concepts/composition-functions.md:343]: Can you add `end-to-end` to 
`utils/vale/styles/Crossplane/spelling-exceptions.txt`

to fix the vale error here?

**Review by plumbis — CHANGES_REQUESTED** [review]
I'm almost done but have a bit more of the compostion-functions page to get through.

File: content/master/concepts/composition-functions.md (hunk 5 above)
Comment (negz) [reply content/master/concepts/composition-functions.md:476]: > this image is hard to read when using dark mode

@JonasKs Yeah, this is a known issue with mermaid graph rendering. We added it in this PR but haven't done any styling yet. I'd like to fix that a subsequent PR as it will probably need @plumbis expertise to do it right.

File: content/master/concepts/composition-functions.md (hunk 6 above)
Comment (negz) [reply content/master/concepts/composition-functions.md]: It was - thanks! No idea what happened there.

File: content/master/concepts/composition-functions.md (hunk 7 above)
Comment (negz) [reply content/master/concepts/composition-functions.md:21]: True, but here I'm building on that to make the point that functions aren't only interesting/useful to folks who want to write their own. I think it's worth it.

File: content/master/concepts/composition-functions.md [hunk 12]
```diff
@@ -1,1026 +1,699 @@
...
-to reconcile a Composite Resource (XR). The preceding example shows a minimal
-`Composition` that uses a Composition Function. Note that it has a `functions`
-array rather than the typical P&T style array of `resources`.
+{{< hint "tip" >}}
+Functions are Crossplane Packages. Read more about Packages in the
+[Packages documentation]({{<ref "packages" >}}).
+{{< /hint >}}
 
-## Enabling functions
+By default, the Function pod installs in the same namespace as Crossplane
```
[the remark this answers is not in the document]
Comment (negz) [reply content/master/concepts/composition-functions.md:53]: No. There's a Crossplane flag `--namespace` that overrides it.

FWIW I copied this section verbatim from the existing Providers documentation. +1 to explaining how this can be overridden but I'd prefer to do that in a separate PR once https://github.com/crossplane/docs/pull/566 has landed, since I imagine there will be more updates.

File: content/master/concepts/composition-functions.md (hunk 10 above)
Comment (negz) [reply content/master/concepts/composition-functions.md:318]: Hrm weird, I didn't notice this warning.

File: content/master/concepts/composition-functions.md (hunk 5 above)
Comment (negz) [reply content/master/concepts/composition-functions.md:477]: Nevermind, looks like @plumbis added a commit that fixes this. 🎉

File: content/master/concepts/composition-functions.md [hunk 13]
```diff
@@ -1,1026 +1,700 @@
...
+
+Composition functions let you replace complicated Compositions with code written
+in your programming language of choice. Crossplane has tools, software
+development kits (SDKs) and templates to help you write a function.
+
+
+<!-- vale write-good.Passive = NO -->
+Here's an example of a tiny, hello world function. This example is written in
+[Go](https://go.dev).
+<!-- vale write-good.Passive = YES -->
```
Comment (plumbis) [inline content/master/concepts/composition-functions.md:372]:
```suggestion
Here's an example of a tiny, hello world function. This example is in
[Go](https://go.dev).
```

**Review by plumbis — APPROVED** [review]
non-blocking nit. Approved!
